### PR TITLE
Clean up dycoms BC

### PR DESCRIPTION
**Pull request by jkozdon**
@mwarusz found a bug in the DYCOMS bc that no boundary conditions was being enforced on the top boundary on the diffusive variables.

This PR fixes that as well as removes unused variables in the boundary conditions (such as some unused moist physics calls) and also refactors the `DYCOMS_BC` to reuse values computed for the `NoFluxBC` whenever possible.

Co-authored-by: Maciej Waruszewski <mwarusz@igf.fuw.edu.pl>

### CODE REVIEW

**Comment by jkozdon** [comment]
bors try

File: src/Atmos/Model/boundaryconditions.jl [hunk 1]
```diff
@@ -228,27 +190,33 @@ function atmos_boundary_flux_diffusive!(nf::NumericalFluxDiffusive,
     # Boundary momentum fluxes
     # ----------------------------------------------------------
     # Case specific for flat bottom topography, normal vector is n⃗ = k⃗ = [0, 0, 1]ᵀ
-    # A more general implementation requires (n⃗ ⋅ ∇A) to be defined where A is replaced by the appropriate flux terms
+    # A more general implementation requires (n⃗ ⋅ ∇A) to be defined where A is
+    # replaced by the appropriate flux terms
     C_drag = bc.C_drag
-    τ13⁺  = - C_drag * windspeed_FN * u_FN
-    τ23⁺  = - C_drag * windspeed_FN * v_FN
+    τ13⁺ = - C_drag * windspeed_FN * u_FN[1]
```
Comment (mwarusz) [inline src/Atmos/Model/boundaryconditions.jl]: Add `@inbounds` here.

File: src/Atmos/Model/boundaryconditions.jl [hunk 2]
```diff
@@ -228,27 +190,33 @@ function atmos_boundary_flux_diffusive!(nf::NumericalFluxDiffusive,
...
+    τ23⁺ = - C_drag * windspeed_FN * u_FN[2]
+    τ21⁺ = τ⁻[2,1]
```
Comment (mwarusz) [inline src/Atmos/Model/boundaryconditions.jl]: Add `@inbounds` here.

Comment (jkozdon) [reply src/Atmos/Model/boundaryconditions.jl]: Good catch. Thanks!

**Comment by smarras79** [comment]
Thanks for that. 
However, the top should only be controlled by the relaxation due to the sponge that should allow a free slip with a no-penetration b.c. on the vertical component on velocity. It may be that a viscous b.c. on the top will actually cause problems for this specific tests. I will test it as soon as we know that the simulation preserves stability all the way to completion.

**Comment by jkozdon** [comment]
Even with a sponge layer, I’m pretty sure you need to use a well-posed boundary condition which it wasn’t before.

**Comment by lcw** [comment]
In my experience with linear wave propagation we always set a boundary condition where the domain is artificially truncated.  This is not a physical boundary condition but is needed to prevent numerical instability.

If the sponge is doing its job, in my experience, it doesn't matter what boundary condition you set for the artificially truncated domain.

**Comment by jkozdon** [comment]
> If the sponge is doing its job, in my experience, it doesn't matter what boundary condition you set for the artificially truncated domain.

But in your experience, we need to set something reasonable (e.g., enough boundary conditions)?

**Comment by lcw** [comment]
> But in your experience, we need to set something reasonable (e.g., enough boundary conditions)?

That is right, something well posed.

**Comment by smarras79** [comment]
> > But in your experience, we need to set something reasonable (e.g., enough boundary conditions)?
> 
> That is right, something well posed.

Perfect, thanks for clarifying. I'll test dycoms with those b.c.
